### DISCUSSION TITLE
Glue in Python

### DISCUSSION THREAD

**Discussion by yohanespm77** [teamalgoritma/community#699] [Q&A] [answered]
Saya sedang membuat plot line chart di python, seperti ini : 
<img width="716" alt="image" src="https://github.com/teamalgoritma/community/assets/106010284/21dc130c-2227-4659-8f1a-1f464407117c">

Namun saya ingin labelnya seperti glue di R studio, sehingga outputnya menjadi seperti ini : 
Attack = ASM Cookie Hijacking
Policy Name = WAF-INTERNET-BANKING
tanggal = Apr 27, 2023
freq = 4605

Mohon bantuannya kak, terima kasih 🙏

**Comment by ramdhinafinita** [comment]
Dear Pak @yohanespm77

Pada plotly python bisa juga kita terapkan hal sama seperti glue() pada R dengan cara mendefinisikan konten tooltip sebagai string dan dimasukkan kedalam attribut `hovertemplate` pada tooltip. 
Sebagai contoh dari hasil output yang diinginkan, yakni pada kolom x yang memiliki 2 informasi dalam satu element text dan ingin dipisahkan tooltipnya bisa menambahakan pada atribut `customdata`. 

Berikut ini contoh nya:

```
import pandas as pd
import plotly.express as px

df = {
    'Attack_Policy': ['ASM Cookie Hijacking - WAF-INTERNET-BANKING', 'ASM 23 - TEXAS', 'ASM 000 - INTERNET'],
    'tanggal': ['Apr 27, 2023', 'Mei 27, 2023', 'Juni 27, 2023'],
    'Freq': [4605, 1234, 3210]
}

df = pd.DataFrame(df)

fig = px.line(df, x='tanggal', y='Freq', color='Attack_Policy', title='Attack Frequency Trend', markers=True)

# Define custom hovertemplate
hovertemplate = (
    'Attack: %{customdata[0]:default}<br>'
    'Policy Name: %{customdata[1]:default}<br>'
    'tanggal: %{x}<br>'
    'freq: %{y}'
)

# Assign customdata and hovertemplate to each trace
for trace in fig.data:
    trace.customdata = [[trace.name.split(' - ')[0], trace.name.split(' - ')[1]]]
    trace.hovertemplate = hovertemplate


fig.update_layout(showlegend=False).show()
```

Untuk referensi lebih lengkap bisa di lihat disini: https://plotly.com/python/hover-text-and-formatting/#hover-labels

**Comment by yohanespm77** [comment]
Terima kasih atas jawabannya kak @ramdhinafinita .

Plotnya jadi seperti ini :

<img width="517" alt="image" src="https://github.com/teamalgoritma/community/assets/106010284/8b1e8ae0-d502-4526-90ed-91bd37747b0f">

Untuk menghilangkan Attack yang di lingkarin itu gimana ya ? Biar ga kedouble attacknya kak

**Reply by ramdhinafinita** [reply]
Dear Pak @yohanespm77 
Pada hovertemplate bisa ditambahkan <extra></extra>

```
hovertemplate = (
    'Attack: %{customdata[0]:default}<br>'
    'Policy Name: %{customdata[1]:default}<br>'
    'tanggal: %{x}<br>'
    'freq: %{y}<br>'
    '<extra></extra>'
)
```

**Reply by yohanespm77** [reply]
Wah iya mantap sekali kak 👍 Makasih banyak kak @ramdhinafinita

**Reply by ramdhinafinita** [reply]
Siap pak.. sama-sama Jika sudah bisa dan solve bisa dibantu untuk mark as answer ya Pak :) Terimakasih

**Reply by yohanespm77** [reply]
Eh sbntar kak, di data saya kok tampilannya jadi seperti ini ya kak ?

<img width="431" alt="image" src="https://github.com/teamalgoritma/community/assets/106010284/5c725d70-b81c-444b-acfb-1e49e01d6ce1">

Tanggal 25 keatas, tampilannya jadi seperti ini : 

<img width="508" alt="image" src="https://github.com/teamalgoritma/community/assets/106010284/c4a117f6-358d-4a84-a78f-ff2f61a2996c">

<img width="598" alt="image" src="https://github.com/teamalgoritma/community/assets/106010284/18947b49-2a4d-422a-82c4-a25c22f2e94a">


Kenapa ya kak ?

**Reply by yohanespm77** [reply]
kodingannya seperti ini kak @ramdhinafinita 


import pandas as pd
import plotly.express as px

df = {
    'tanggal': ['2023-04-24', '2023-04-24', '2023-04-24','2023-04-25','2023-04-25','2023-04-25','2023-04-26','2023-04-26','2023-04-26'],
    'Attack': ['ASM 23', 'ASM 23', 'ASM 00','Data Guard','ASM 00','ASM 00','ASM 00','Illegal HTTP status','ASM 23'],
    'Policy Name': ['INTERNET','OMNI','TEXAS','OPEN_API','LINK','TEXAS','TEXAS','WAF_E-OMNI','INTERNET'],
    'Freq':[433,35,75,6,3,111,123,56,100],
    'Attack_Policy':['ASM 23 | INTERNET','ASM 23 | OMNI', 'Access from disallowed Geolocation | TEXAS', 'Data Guard | INTERNET', 'ASM 00 | LINK', 'Illegal URL | TEXAS','Access from disallowed Geolocation | TEXAS', 'ASM 23 | OMNI', 'ASM 23 | INTERNET']
}

df = pd.DataFrame(df)

df['tanggal'] = pd.to_datetime(df['tanggal'])

fig = px.line(df, x='tanggal', y='Freq', color='Attack_Policy', title='Attack Frequency Trend', markers=True)

hovertemplate = (
    'Attack: %{customdata[0]}<br>'
    'Policy Name: %{customdata[1]}<br>'
    'tanggal: %{x|%Y-%m-%d}<br>'
    'freq: %{y}'
    '<extra></extra>'
)

for trace in fig.data:
    trace.customdata = [[trace.name.split(' | ')[0], trace.name.split(' | ')[1]]]
    trace.hovertemplate = hovertemplate

fig.update_layout(showlegend=False)

fig.show()

**Comment by ramdhinafinita** [comment, marked as answer]
Dear pak @yohanespm77 

kode yang bapak tulis seharusnya pada looping trace disesuaikan pada bagian customdatanya.

berikut contoh nya:

```
for trace in fig.data:
    attack, policy = trace.name.split(' | ')
    customdata = [[attack, policy]] * len(df)
    trace.update(customdata=customdata, hovertemplate=hovertemplate)
```

sehingga outputnya seharusnya sudah berhasil loop
![image](https://github.com/teamalgoritma/community/assets/64061969/6625dbd5-1cb5-4800-bfe9-41e0495a6460)

semoga membantu

**Reply by yohanespm77** [reply]
Baik terima kasih kak @ramdhinafinita
